File: flutter/lib/core/parsers.dart
import 'dart:convert';
import 'types.dart';

/// Defensive parsers — the Dart mirror of `packages/shared/src/parsers.ts`.
/// Each degrades gracefully when JSON is absent.

dynamic _tryJson(String text) {
  final trimmed = text.trim();
  if (trimmed.isEmpty) return null;
  final start = trimmed.indexOf(RegExp(r'[\[{]'));
  if (start == -1) return null;
  try {
    return jsonDecode(trimmed.substring(start));
  } catch (_) {
    return null;
  }
}

num? _num(dynamic v) => v is num ? v : null;
String? _str(dynamic v) => v is String ? v : null;

/// Surge may wrap payloads as {result, error, <payload>}. Unwrap by key.
dynamic _unwrap(Map outer, String key) => outer.containsKey(key) ? outer[key] : outer;

/// Parse `ls -1` of the config dir into profile names (`*.conf`, sans extension).
List<String> parseProfiles(String stdout) {
  final out = stdout
      .split(RegExp(r'\r?\n'))
      .map((l) => l.trim())
      .where((l) => l.endsWith('.conf'))
      .map((l) => l.replaceFirst(RegExp(r'\.conf$'), ''))
      .toList()
    ..sort();
  return out;
}

/// Surface a Surge `error` field from a --raw response, if any.
String? extractError(String stdout) {
  final json = _tryJson(stdout);
  return json is Map ? _str(json['error']) : null;
}

Environment parseEnvironment(String stdout) {
  final json = _tryJson(stdout);
  final fields = <String, String>{};
  final selection = <String, String>{};
  final autoOverride = <String, String>{};
  String? globalPolicy;
  int? proxyMode;

  final env = json is Map ? _unwrap(json, 'environment') : null;
  if (env is Map) {
    env.forEach((k, v) {
      if (v == null) return;
      if (k == 'ProxyGroupSelection' && v is Map) {
        v.forEach((g, p) {
          final name = _str(p);
          if (name != null) selection['$g'] = name;
        });
        return;
      }
      if (k == 'AutoPolicyGroupOverride' && v is Map) {
        v.forEach((g, p) {
          final name = _str(p);
          if (name != null) autoOverride['$g'] = name;
        });
        return;
      }
      if (k == 'ProxyMode') {
        proxyMode = _num(v)?.toInt() ?? int.tryParse('$v');
      }
      if (k == 'AllProxyModePolicyNameKey') {
        globalPolicy = _str(v);
      }
      fields['$k'] = v is Map || v is List ? jsonEncode(v) : '$v';
    });
  }
  return Environment(
    fields: fields,
    selection: selection,
    autoOverride: autoOverride,
    globalPolicy: globalPolicy,
    proxyMode: proxyMode,
    raw: json ?? stdout,
  );
}

/// `surge --raw dump policy-group-sub-policies` → group → member names.
Map<String, List<String>> parseSubPolicies(String stdout) {
  final json = _tryJson(stdout);
  if (json is! Map) return {};
  final rec = _unwrap(json, 'policy-group-sub-policies');
  final map = rec is Map ? rec : json;
  final out = <String, List<String>>{};
  map.forEach((group, value) {
    if (value is List) {
      out['$group'] = _names(value);
    } else if (value is Map) {
      out['$group'] = _names(value['all'] ?? value['members'] ?? value['subPolicies']);
    }
  });
  return out;
}

String? _nameOf(dynamic v) {
  if (v is String) return v;
  if (v is Map) {
    return _str(v['name']) ?? _str(v['key']) ?? _str(v['url']) ?? _str(v['path']);
  }
  return null;
}

List<String> _names(dynamic v) => v is List
    ? v.map(_nameOf).whereType<String>().toList()
    : const [];

/// `surge --raw dump policy` → {"proxies":[...],"policy-groups":[...]}
PolicyDump parsePolicies(String stdout) {
  final json = _tryJson(stdout);
  if (json is Map) {
    return PolicyDump(
      proxies: _names(json['proxies']),
      groups: _names(json['policy-groups'] ?? json['policyGroups'] ?? json['groups']),
    );
  }
  return const PolicyDump();
}

/// `surge --raw dump smart-group-info` → smart group names keyed by group.
Map<String, String> parseSmartGroupTypes(String stdout) {
  final json = _tryJson(stdout);
  if (json is! Map) return const {};
  final out = <String, String>{};
  json.forEach((name, value) {
    if (name == 'report') return;
    if (value is Map) out['$name'] = 'smart';
  });
  return out;
}

/// `surge --raw test-all-policies` → {"UK":{"tcp":66,"receive":415,...}, ...}
List<PolicyTest> parsePolicyTests(String stdout) {
  final json = _tryJson(stdout);
  if (json is! Map) return const [];
  final out = <PolicyTest>[];
  json.forEach((name, value) {
    if (value is! Map) return;
    out.add(PolicyTest(
      name: '$name',
      tcpMs: _num(value['tcp'])?.toInt(),
      receiveMs: _num(value['receive'])?.toInt(),
      available: _num(value['available'])?.toInt(),
      roundOneTotal: _num(value['round-one-total'])?.toInt(),
      error: _str(value['error']),
    ));
  });
  return out;
}

/// Parse one classic comma-separated rule line into a Rule.
///   "DOMAIN-SUFFIX,google.com,Proxy" → matcher rule (type,value,policy)
///   "FINAL,Proxy" / "FINAL,Proxy,dns-failed" → FINAL rule (no value)
Rule? _parseRuleLine(String line) {
  final parts = line.split(',').map((p) => p.trim()).toList();
  if (parts.isEmpty) return null;
  final type = parts[0];
  if (type.isEmpty) return null;
  // FINAL has no matcher value: the second token is the policy itself.
  if (type == 'FINAL') {
    return parts.length >= 2 ? Rule(type: type, value: '', policy: parts[1]) : null;
  }
  if (parts.length < 3) return null;
  return Rule(type: type, value: parts[1], policy: parts[2]);
}

/// `surge --raw dump rule` → {"rules":["DOMAIN-SUFFIX,google.com,Proxy", ...]}
/// (each entry is a rule string). Tolerant of a bare array, an array of
/// {type,value,policy} objects, or plain newline-delimited text.
List<Rule> parseRules(String stdout) {
  final json = _tryJson(stdout);
  final arr = json is List
      ? json
      : (json is Map && json['rules'] is List)
          ? json['rules'] as List
          : null;
  if (arr != null) {
    final out = <Rule>[];
    for (final item in arr) {
      if (item is String) {
        final r = _parseRuleLine(item);
        if (r != null) out.add(r);
      } else if (item is Map) {
        final type = _str(item['type']) ?? '';
        if (type.isEmpty) continue;
        out.add(Rule(
          type: type,
          value: _str(item['value']) ?? _str(item['pattern']) ?? '',
          policy: _str(item['policy']) ?? _str(item['target']) ?? '',
          hits: _num(item['hits'])?.toInt() ?? _num(item['count'])?.toInt(),
        ));
      }
    }
    return out;
  }
  // Fallback: classic comma-separated rule lines, one per line.
  final out = <Rule>[];
  for (final line in const LineSplitter().convert(stdout)) {
    final t = line.trim();
    if (t.isEmpty || t.startsWith('#')) continue;
    final r = _parseRuleLine(t);
    if (r != null) out.add(r);
  }
  return out;
}

/// `surge --raw dump temp-rule` → raw rule strings (what del-temp-rule expects).
List<String> parseTempRules(String stdout) {
  final json = _tryJson(stdout);
  final arr = json is List
      ? json
      : (json is Map && json['temp-rule'] is List)
          ? json['temp-rule'] as List
          : null;
  if (arr != null) {
    return arr
        .map((item) {
          if (item is String) return item;
          if (item is Map) {
            return [
              _str(item['type']),
              _str(item['value']) ?? _str(item['pattern']),
              _str(item['policy']),
            ].whereType<String>().join(',');
          }
          return '';
        })
        .where((s) => s.isNotEmpty)
        .toList();
  }
  return stdout
      .split(RegExp(r'\r?\n'))
      .map((l) => l.trim())
      .where((l) => l.isNotEmpty && !l.startsWith('#'))
      .toList();
}

/// `surge --raw external-resource list`.
List<ExternalResource> parseExternalResources(String stdout) {
  final json = _tryJson(stdout);
  final list = json is List
      ? json
      : (json is Map && json['resources'] is List)
          ? json['resources'] as List
          : const [];
  return list.whereType<Map>().map((r) {
    return ExternalResource(
      key: _str(r['key']) ?? _str(r['hash']) ?? _str(r['url']) ?? '',
      url: _str(r['url']),
      ready: r['ready'] is bool ? r['ready'] as bool : null,
      updatedAt: _num(r['updatedAt'])?.toInt() ?? _num(r['updated'])?.toInt(),
    );
  }).where((r) => r.key.isNotEmpty).toList();
}

/// Coerce a string-or-number id into a string (Surge uses numeric ids).
String? _idStr(dynamic v) {
  if (v is num) return '$v';
  return _str(v);
}

/// `surge --raw dump active` → list of active connections. Surge returns the
/// active requests under a `requests` envelope (`{"requests":[...]}`); older/
/// forked shapes use `connections` or a bare array — all are accepted.
List<ActiveConnection> parseActive(String stdout) {
  final json = _tryJson(stdout);
  final list = json is List
      ? json
      : (json is Map && json['requests'] is List)
          ? json['requests'] as List
          : (json is Map && json['connections'] is List)
              ? json['connections'] as List
              : const [];
  var i = 0;
  return list.whereType<Map>().map((c) {
    final conn = ActiveConnection(
      // Surge request ids are numbers; keep them as strings for `kill <id>`.
      id: _idStr(c['id']) ?? _idStr(c['connectionId']) ?? '$i',
      remote: _str(c['remoteAddress']) ??
          _str(c['remote']) ??
          _str(c['host']) ??
          _str(c['URL']) ??
          _str(c['url']) ??
          '—',
      policy: _str(c['policyName']) ?? _str(c['policy']) ?? _str(c['proxy']),
      rule: _str(c['rule']),
      uploadBytes: _num(c['outBytes']) ?? _num(c['uploadBytes']) ?? _num(c['upload']),
      downloadBytes: _num(c['inBytes']) ?? _num(c['downloadBytes']) ?? _num(c['download']),
    );
    i++;
    return conn;
  }).toList();
}

/// Aggregate `dump active` connections into dashboard totals.
Traffic aggregateTraffic(List<ActiveConnection> connections) {
  num up = 0;
  num down = 0;
  for (final c in connections) {
    up += c.uploadBytes ?? 0;
    down += c.downloadBytes ?? 0;
  }
  return Traffic(
    connections: connections.length,
    uploadTotal: up == 0 ? null : up,
    downloadTotal: down == 0 ? null : down,
  );
}

const _levels = [
  LogLevel.debug,
  LogLevel.info,
  LogLevel.notify,
  LogLevel.warning,
  LogLevel.error,
];

LogLine parseLogLine(String line) {
  final lower = line.toLowerCase();
  final level = _levels.firstWhere(
    (l) => lower.contains('[${l.name}]') || lower.contains(' ${l.name} '),
    orElse: () => LogLevel.unknown,
  );
  final tsMatch = RegExp(r'^\[?(\d{4}-\d{2}-\d{2}[ T][\d:.]+)').firstMatch(line);
  final ts = tsMatch != null
      ? (DateTime.tryParse(tsMatch.group(1)!)?.millisecondsSinceEpoch ??
          DateTime.now().millisecondsSinceEpoch)
      : DateTime.now().millisecondsSinceEpoch;
  return LogLine(ts: ts, level: level, message: line);
}

String formatBps(num? bps) {
  if (bps == null || bps <= 0) return '0 B/s';
  const units = ['B/s', 'KB/s', 'MB/s', 'GB/s'];
  var v = bps.toDouble();
  var i = 0;
  while (v >= 1024 && i < units.length - 1) {
    v /= 1024;
    i++;
  }
  final digits = (v >= 100 || i == 0) ? 0 : 1;
  return '${v.toStringAsFixed(digits)} ${units[i]}';
}

String formatBytes(num? bytes) {
  if (bytes == null || bytes <= 0) return '—';
  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  var v = bytes.toDouble();
  var i = 0;
  while (v >= 1024 && i < units.length - 1) {
    v /= 1024;
    i++;
  }
  return '${v.toStringAsFixed(i == 0 ? 0 : 1)} ${units[i]}';
}

File: packages/shared/src/parsers.ts
import type {
  ActiveConnection,
  Environment,
  ExternalResource,
  LogLine,
  PolicyDump,
  PolicyTest,
  Rule,
  Traffic,
} from "./types.js";

/**
 * Parsers turn raw `surge` stdout into domain models. They are intentionally
 * defensive: surge forks and versions differ, JSON may be absent, so each parser
 * degrades gracefully and keeps the raw payload for debugging.
 */

function tryJson(text: string): unknown {
  const trimmed = text.trim();
  if (!trimmed) return undefined;
  // Surge may print a banner line before JSON; grab the first {...} or [...].
  const start = trimmed.search(/[[{]/);
  if (start === -1) return undefined;
  try {
    return JSON.parse(trimmed.slice(start));
  } catch {
    return undefined;
  }
}

function asRecord(v: unknown): Record<string, unknown> | undefined {
  return v && typeof v === "object" && !Array.isArray(v)
    ? (v as Record<string, unknown>)
    : undefined;
}

/**
 * Surge sometimes wraps payloads in a response envelope
 * (`{ "result": ..., "error": ..., "<payload>": ... }`). Given a payload key,
 * return the inner payload when present, else the object itself.
 */
function unwrap(obj: Record<string, unknown>, key: string): unknown {
  if (key in obj) return obj[key];
  return obj;
}

/**
 * Parse `ls -1` output of the config directory into profile names: keep only
 * `*.conf` files and strip the extension (the name `switch-profile` expects).
 */
export function parseProfiles(stdout: string): string[] {
  return stdout
    .split(/\r?\n/)
    .map((l) => l.trim())
    .filter((l) => l.endsWith(".conf"))
    .map((l) => l.replace(/\.conf$/, ""))
    .sort((a, b) => a.localeCompare(b));
}

/** Surface a Surge `error` field from a --raw response, if any. */
export function extractError(stdout: string): string | undefined {
  const rec = asRecord(tryJson(stdout));
  return rec ? str(rec.error) : undefined;
}

function num(v: unknown): number | undefined {
  return typeof v === "number" && Number.isFinite(v) ? v : undefined;
}

function str(v: unknown): string | undefined {
  return typeof v === "string" ? v : undefined;
}

/** Flatten `surge --raw environment` JSON into displayable key/value pairs. */
export function parseEnvironment(stdout: string): Environment {
  const json = tryJson(stdout);
  const fields: Record<string, string> = {};
  const selection: Record<string, string> = {};
  let proxyMode: number | undefined;

  const outer = asRecord(json);
  const env = asRecord(outer ? unwrap(outer, "environment") : undefined);
  if (env) {
    for (const [k, v] of Object.entries(env)) {
      if (v === null || v === undefined) continue;
      if (k === "ProxyGroupSelection") {
        const sel = asRecord(v);
        if (sel) {
          for (const [g, p] of Object.entries(sel)) {
            const name = str(p);
            if (name) selection[g] = name;
          }
        }
        continue;
      }
      if (k === "ProxyMode") {
        proxyMode = num(v) ?? Number(v);
        if (!Number.isFinite(proxyMode)) proxyMode = undefined;
      }
      fields[k] = typeof v === "object" ? JSON.stringify(v) : String(v);
    }
  }
  return { fields, selection, proxyMode, raw: json ?? stdout };
}

function nameOf(v: unknown): string | undefined {
  if (typeof v === "string") return v;
  const rec = asRecord(v);
  return rec ? str(rec.name) ?? str(rec.key) ?? str(rec.url) ?? str(rec.path) : undefined;
}

function toNames(v: unknown): string[] {
  return Array.isArray(v)
    ? v.map(nameOf).filter((m): m is string => !!m)
    : [];
}

/**
 * `surge --raw dump policy` →
 *   {"proxies":["UK","US",...],"policy-groups":["Relay","Apple",...]}
 */
export function parsePolicies(stdout: string): PolicyDump {
  const rec = asRecord(tryJson(stdout));
  return {
    proxies: toNames(rec?.proxies),
    groups: toNames(rec?.["policy-groups"] ?? rec?.policyGroups ?? rec?.groups),
  };
}

/**
 * `surge --raw dump policy-group-sub-policies` → group → member names.
 * Tolerant of `{ "Group": [...] }` and `{ "Group": { "all": [...] } }`.
 */
export function parseSubPolicies(stdout: string): Record<string, string[]> {
  const outer = asRecord(tryJson(stdout));
  if (!outer) return {};
  const rec =
    asRecord(outer.map) ??
    asRecord(unwrap(outer, "policy-group-sub-policies")) ??
    outer;
  const out: Record<string, string[]> = {};
  for (const [group, value] of Object.entries(rec)) {
    if (Array.isArray(value)) out[group] = toNames(value);
    else {
      const g = asRecord(value);
      if (g) out[group] = toNames(g.all ?? g.members ?? g.subPolicies);
    }
  }
  return out;
}

/**
 * `surge --raw test-all-policies` / `test-policy` / `test-group` →
 *   {"UK":{"tcp":66,"receive":415,"available":69,"round-one-total":1055},
 *    "CA":{"error":"Socket closed by remote peer","available":0}, ...}
 */
export function parsePolicyTests(stdout: string): PolicyTest[] {
  const rec = asRecord(tryJson(stdout));
  if (!rec) return [];
  const out: PolicyTest[] = [];
  for (const [name, value] of Object.entries(rec)) {
    const r = asRecord(value);
    if (!r) continue;
    out.push({
      name,
      tcpMs: num(r.tcp),
      receiveMs: num(r.receive),
      available: num(r.available),
      roundOneTotal: num(r["round-one-total"]),
      error: str(r.error),
    });
  }
  return out;
}


/**
 * Parse one classic comma-separated rule line into a Rule.
 *   "DOMAIN-SUFFIX,google.com,Proxy"  → matcher rule (type,value,policy)
 *   "FINAL,Proxy"  /  "FINAL,Proxy,dns-failed"  → FINAL rule (no value)
 * Trailing options (e.g. `no-resolve`, `dns-failed`) after the policy are kept
 * out of the model.
 */
function parseRuleLine(line: string): Rule | undefined {
  const parts = line.split(",").map((p) => p.trim());
  const type = parts[0];
  if (!type) return undefined;
  // FINAL has no matcher value: the second token is the policy itself.
  if (type === "FINAL") {
    return parts[1] ? { type, value: "", policy: parts[1] } : undefined;
  }
  if (parts.length < 3) return undefined;
  return { type, value: parts[1]!, policy: parts[2]! };
}

/**
 * `surge --raw dump rule` → `{"rules":["DOMAIN-SUFFIX,google.com,Proxy", ...]}`
 * (each entry is a classic rule string). Tolerant of a bare array, an array of
 * `{type,value,policy}` objects, or plain newline-delimited text.
 */
export function parseRules(stdout: string): Rule[] {
  const json = tryJson(stdout);
  const arr = ruleArray(json);
  if (arr) {
    return arr
      .map((item): Rule | undefined => {
        if (typeof item === "string") return parseRuleLine(item);
        const r = asRecord(item);
        if (!r) return undefined;
        const type = str(r.type) ?? "";
        if (!type) return undefined;
        return {
          type,
          value: str(r.value) ?? str(r.pattern) ?? "",
          policy: str(r.policy) ?? str(r.target) ?? "",
          hits: num(r.hits) ?? num(r.count),
        };
      })
      .filter((r): r is Rule => !!r);
  }
  // Fallback: classic comma-separated rule lines, one per line.
  return stdout
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line && !line.startsWith("#"))
    .map(parseRuleLine)
    .filter((r): r is Rule => !!r);
}

/** Extract the rule array from a bare array or a `{ "rules": [...] }` envelope. */
function ruleArray(json: unknown): unknown[] | undefined {
  if (Array.isArray(json)) return json;
  const rec = asRecord(json);
  if (rec && Array.isArray(rec.rules)) return rec.rules as unknown[];
  return undefined;
}

/**
 * Parse `surge --raw dump temp-rule` into raw rule strings. Each string is the
 * exact rule line that `del-temp-rule <rule>` expects back. Tolerant of an
 * array of strings, an array of {type,value,policy} objects, or plain text.
 */
export function parseTempRules(stdout: string): string[] {
  const json = tryJson(stdout);
  const arr = Array.isArray(json)
    ? json
    : Array.isArray(asRecord(json)?.["temp-rule"])
      ? (asRecord(json)!["temp-rule"] as unknown[])
      : Array.isArray(asRecord(json)?.rules)
        ? (asRecord(json)!.rules as unknown[])
      : undefined;
  if (arr) {
    return arr
      .map((item) => {
        if (typeof item === "string") return item;
        const r = asRecord(item);
        if (!r) return "";
        const parts = [str(r.type), str(r.value) ?? str(r.pattern), str(r.policy)]
          .filter((p): p is string => !!p);
        return parts.join(",");
      })
      .filter((s) => s);
  }
  return stdout
    .split(/\r?\n/)
    .map((l) => l.trim())
    .filter((l) => l && !l.startsWith("#"));
}

/** Parse `surge --raw external-resource list`. */
export function parseExternalResources(stdout: string): ExternalResource[] {
  const json = tryJson(stdout);
  const list = Array.isArray(json)
    ? json
    : Array.isArray(asRecord(json)?.resources)
      ? (asRecord(json)!.resources as unknown[])
      : Array.isArray(asRecord(json)?.defines)
        ? (asRecord(json)!.defines as unknown[])
      : [];
  return list
    .map(asRecord)
    .filter((r): r is Record<string, unknown> => !!r)
    .map((r) => ({
      key: str(r.key) ?? str(r.hash) ?? str(r.url) ?? "",
      url: str(r.url) ?? str(r.path),
      ready: typeof r.ready === "boolean" ? r.ready : undefined,
      updatedAt: num(r.updatedAt) ?? num(r.updated),
    }))
    .filter((r) => r.key);
}

/**
 * Parse `surge --raw dump active` into a connection list. Surge returns the
 * active requests under a `requests` envelope (`{"requests":[...]}`); older/forked
 * shapes use `connections` or a bare array — all are accepted.
 */
export function parseActive(stdout: string): ActiveConnection[] {
  const json = tryJson(stdout);
  const rec = asRecord(json);
  const list = Array.isArray(json)
    ? json
    : Array.isArray(rec?.requests)
      ? (rec!.requests as unknown[])
      : Array.isArray(rec?.["active-requests"])
        ? (rec!["active-requests"] as unknown[])
      : Array.isArray(rec?.connections)
        ? (rec!.connections as unknown[])
        : [];
  return list
    .map(asRecord)
    .filter((c): c is Record<string, unknown> => !!c)
    .map((c, i) => ({
      // Surge request ids are numbers; keep them as strings for `kill <id>`.
      id: idStr(c.id) ?? idStr(c.connectionId) ?? String(i),
      remote:
        str(c.remoteHost) ??
        str(c.URL) ??
        str(c.remoteAddress) ??
        str(c.remote) ??
        str(c.host) ??
        str(c.URL) ??
        str(c.url) ??
        "—",
      policy:
        str(c.policyName) ??
        str(c.originalPolicyName) ??
        str(c.policy) ??
        str(c.proxy),
      rule: str(c.rule) ?? str(c.ruleName),
      uploadBytes: num(c.outBytes) ?? num(c.uploadBytes) ?? num(c.upload),
      downloadBytes: num(c.inBytes) ?? num(c.downloadBytes) ?? num(c.download),
      raw: c,
    }));
}

/** Coerce a string-or-number id into a string (Surge uses numeric ids). */
function idStr(v: unknown): string | undefined {
  if (typeof v === "number" && Number.isFinite(v)) return String(v);
  return str(v);
}

/** Aggregate `dump active` connections into totals for the dashboard. */
export function aggregateTraffic(connections: ActiveConnection[]): Traffic {
  let uploadTotal = 0;
  let downloadTotal = 0;
  for (const c of connections) {
    uploadTotal += c.uploadBytes ?? 0;
    downloadTotal += c.downloadBytes ?? 0;
  }
  return {
    connections: connections.length,
    uploadTotal: uploadTotal || undefined,
    downloadTotal: downloadTotal || undefined,
  };
}

const LOG_LEVELS = ["debug", "info", "notify", "warning", "error"] as const;

export function parseLogLine(line: string): LogLine {
  const lower = line.toLowerCase();
  const level =
    LOG_LEVELS.find((l) => lower.includes(`[${l}]`) || lower.includes(` ${l} `)) ??
    "unknown";
  // Try to extract a leading ISO/epoch timestamp; otherwise stamp now.
  const tsMatch = line.match(/^\[?(\d{4}-\d{2}-\d{2}[ T][\d:.]+)/);
  const ts = tsMatch ? Date.parse(tsMatch[1]!) || Date.now() : Date.now();
  return { ts, level, message: line };
}

/** Human-readable bytes/sec. */
export function formatBps(bps?: number): string {
  if (!bps || bps <= 0) return "0 B/s";
  const units = ["B/s", "KB/s", "MB/s", "GB/s"];
  let v = bps;
  let i = 0;
  while (v >= 1024 && i < units.length - 1) {
    v /= 1024;
    i++;
  }
  return `${v.toFixed(v >= 100 || i === 0 ? 0 : 1)} ${units[i]}`;
}
